gtsam/sfm/ShonanAveraging.cpp

```diff
@@ -338,11 +336,23 @@ template <typename T>
 static double Kappa(const BinaryMeasurement<T> &measurement) {
   const auto &isotropic = boost::dynamic_pointer_cast<noiseModel::Isotropic>(
       measurement.noiseModel());
-  if (!isotropic) {
-    throw std::invalid_argument(
-        "Shonan averaging noise models must be isotropic.");
+  double sigma;
+  if (isotropic) {
+    sigma = isotropic->sigma();
+  } else {
+    const auto &robust = boost::dynamic_pointer_cast<noiseModel::Robust>(
+        measurement.noiseModel());
+    if (robust) {
+      std::cout << "Verification of optimality does not work with robust cost "
```

Comment:
It should throw.

Reply:
This causes the constructor to error out when using the huber noise model. Thoughts?

Reply:
Also, I just noticed that the else condition doesn't assign a value to sigma. Updating it.

Reply:
Actually this whole if-else doesn't make any sense to me. If the noise model is required to be isotropic, shouldn't we also be checking if the robust model's noise model is isotropic?

Reply:
Try remove the autos?

Reply:
I don't think that will help, because there are test cases where we specifically use the huber model and this if-case will succeed. Throwing an exception will cause the test to fail.

Reply:
You can test an exception, no? You can't check whether something prints out. Anyway, printing a warning is not the way to handle a wrong use of the module.

Reply:
I am not sure I understand. Shonan should still be able to work with a robust model, only losing the power to provide the certificate of optimality, right?

Reply:
There is a ‘robust’ flag somewhere, right? If the flag is set, then robust should be allowed without a message. If it is not set it should throw. Printing out a message and then muddling on is not an option in my view.

Reply:
Gotcha. I did misunderstand what you meant, sorry about that.

Reply:
my 2 cents: it might be fishy to just disable the optimality verification when using robust loss, since the user might still think s/he is getting a certifiably-optimal solution. I suggest adding a flag "certifyOptimality" and throw an exception when certifyOptimality = true and we use a robust loss. The flag certifyOptimality might be useful also in the non-robust case since the user might want to disable the certification to save time/computation.